### PR TITLE
feat(node): v11.10

### LINKED ISSUE #32965: Null initialization vectors in Node crypto module
- [x] I tried using the `@types/node` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @SimonSchick 

In createCipheriv and createDecipheriv, the initialization vector should be allowed to be null as described in https://nodejs.org/api/crypto.html#crypto_crypto_createcipheriv_algorithm_key_iv_options and https://nodejs.org/api/crypto.html#crypto_crypto_createdecipheriv_algorithm_key_iv_options ("If the cipher does not need an initialization vector, iv may be null.")

Currently, something like createCipheriv("aes-256-ecb", key, null) fails on the third parameter, and it's valid code for using AES in ECB mode (which does not use an IV).

### LINKED ISSUE #33125: Incorrect type for @types/node ChildProcess.stdio
- [x] I tried using the `@types/xxxx` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @SimonSchick, @rbuckton 

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/5f59f3dc7a0965fb1767cf13c670c633bbbbf0a8/types/node/child_process.d.ts#L10 maps `ChildProcess.stdio` to a 3-tuple, but per https://nodejs.org/api/child_process.html#child_process_subprocess_stdio it's an array that can have more than three elements.

### PR DESCRIPTION
**Pull request by SimonSchick**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: 
  - https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V11.md#11.10.0 
  - https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V11.md#2019-01-17-version-1170-current-bridgear
  - https://nodejs.org/docs/latest-v11.x/api/child_process.html#child_process_options_stdio
- [x] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

Added child process `connected` property.
Made all `childProcess` properties readonly (aside from std*).
Made `childProcess.stdio` an array rather than a tuple.
Made `create(De)Cipher` IV nullable.
Added `process.features`.
Added `http(s|2)` `writeHead` return value.
Added `perf_hooks.monitorEventLoopDelay`.
Added `replServer.setupHistory`.
Added `tls` `session` event for socket.
Added `brotli` to crypto.
Generally moved affected tests to `/test` folder

Fixes #33125
Fixes #32965

### CODE REVIEW

**Comment by SimonSchick** [comment]
@Flarna :)

**Comment by SimonSchick** [comment]
Hey @rbuckton can you merge this?
CI failures are unrelated :)

File: types/node/zlib.d.ts
Comment (Flarna) [inline types/node/zlib.d.ts]: Maybe we should name the properties here and use the correct type per option? Not sure if this will work if users use the constants provided.

Comment (SimonSchick) [reply types/node/zlib.d.ts]: Tried that, TS 2.1 says no, especially since I can't refer to the `constants` this would be a giant mess, if you can figure out how to make it work in a maintainable manner I'll gladly add it.

File: types/node/child_process.d.ts
Comment (Flarna) [inline types/node/child_process.d.ts]: Why the change of `stdio`? I think [1] and [2] should be kept as `stream.Readable | null`.

Comment (SimonSchick) [reply types/node/child_process.d.ts]: Oh yea, this is correct, is there a way to declare the first few elements as one type and the `rest` as others?

File: types/node/child_process.d.ts
Comment (Flarna) [inline types/node/child_process.d.ts]: I don't think `channel` is a `boolean`. I player around and got `null` and `undefined` and according to the docs it's a pipe.

File: types/node/child_process.d.ts
Comment (SimonSchick) [reply types/node/child_process.d.ts]: I can do `        readonly stdio: [stream.Writable | null, stream.Readable | null, stream.Readable | null, ...Array<stream.Readable | stream.Writable | null>];
` for TS 3.x, but not for TS 2.x

File: types/node/child_process.d.ts
Comment (SimonSchick) [reply types/node/child_process.d.ts]: Yea my bad, any idea what sort of `Pipe` it is, is that a Duplex stream or something? The NodeJS code is extremely hard to follow here.

File: types/node/child_process.d.ts
Comment (Flarna) [reply types/node/child_process.d.ts]: Don't know if this is possible with <3.x.
You could just add one or two `stream.Readable | stream.Writable | null` elements which is not fully correct but will most likely fit for most users. Seems noone missed them till now anyway...

File: types/node/child_process.d.ts
Comment (Flarna) [reply types/node/child_process.d.ts]: I think it's a `pipe_wrap` which is also used inside a `net.socket` if it is connected via a unix domain socket or windows pipe. Not sure if there exists a detailed doc or an existing type definition.

Comment (SimonSchick) [reply types/node/child_process.d.ts]: We could create a bare-bone `Pipe` interface in the `stream` module, not sure if we are reaching too far into nodes core with that tho.

Comment (Flarna) [reply types/node/child_process.d.ts]: Are you using this feature currently? Maybe we should wait until someone really needs this to have a real world test.

Comment (SimonSchick) [reply types/node/child_process.d.ts]: Nah, but I just wanted to add it for the sake of completeness, I will make it an empty interface for now.

**Review by Flarna — APPROVED** [review]

**Review by BrunoScheufler — APPROVED** [review]

**Comment by SimonSchick** [comment]
@amcasey can we get some love please? ❤️

**Comment by amcasey** [comment]
@SimonSchick I'm seeing a CI failure.  Is it unrelated?

**Comment by Flarna** [comment]
@amcasey CI fails because of some issue in typescript@next, see https://github.com/Microsoft/TypeScript/issues/29898
I would say that >90% of CI failures in node are unrelated as changes from dtslint, typescript,... are introduced here without running CI before.

**Comment by amcasey** [comment]
Closing and re-opening to retrigger CI.

**Comment by Flarna** [comment]
@SimonSchick Seems there is a real lint error inbetween the unrelated errors:
```
Error in node
Error: /home/travis/build/DefinitelyTyped/DefinitelyTyped/types/node/child_process.d.ts:87:1
ERROR: 87:1  no-consecutive-blank-lines  Consecutive blank lines are forbidden
```

**Comment by SimonSchick** [comment]
@Flarna should be fixed 😄

**Comment by SimonSchick** [comment]
@amcasey node CI is still kinda broken, build failures unrelated to this PR.

**Comment by amcasey** [comment]
@SimonSchick I haven't been seeing problems with other PRs, but I'll take a look.

**Comment by amcasey** [comment]
Okay, it turns out we recently enabled some lint rules that were incorrectly not being run and that broke a bunch of packages.  Our preference would be to wait until that is sorted out so we can get a green CI run before merging this.  However, if this is a particularly high value and high confidence change, then I'm willing to bypass the CI.  Thoughts?

**Comment by Flarna** [comment]
> Our preference would be to wait until that is sorted out so we can get a green CI run before merging this.

@amcasey Any timeline on this? or pointers to PRs we can monitor?

I think we should at least merge #33098. Depending on the timeline for getting a green CI also other PRs. Till CI is fixed Bot should not longer close PRs with failed CI.

If you search for PRs closed by @typescript-bot during the last days you will see several ones for node... Sooner or later people will stop submitting PRs here if they get anyway closed by bot and they can't to anything to avoid this...

**Comment by amcasey** [comment]
@Flarna You can monitor the issue in this build queue: https://typescript.visualstudio.com/TypeScript/_build?definitionId=15 or ping @sandersn if you have an urgent question.

I'll have a look at the closed PRs.

**Comment by SimonSchick** [comment]
@amcasey can we merge this regardless?

**Comment by amcasey** [comment]
@SimonSchick, it looks like the fix is still in progress.  I'd rather not merge a change to a package as important as node without a green CI run unless there's something that makes it urgent.  @Flarna suggested that there was a case for bypassing CI for #33098 and that change has been merged.

**Comment by SimonSchick** [comment]
I just don't want the bot to close this PR again is all...

**Comment by SimonSchick** [comment]
Furthermore, the CI has a record of breaking for @types/node every other day, since pretty much every single package depends on it.
Every time something in typescript@next breaks, all PRs get blocked, I've seen numerous PRs fall through the cracks and being re-opened/created multiple times due to this issue and the bot aggressively closing them, it's getting kinda frustrating.

**Comment by sandersn** [comment]
@SimonSchick I'm trying to get Definitely Typed to run cleanly and stay clean on typescript@next, [starting with a daily build](https://typescript.visualstudio.com/TypeScript/_build?definitionId=15&_a=summary).

We're discussing the process at https://github.com/Microsoft/TypeScript/issues/29898. When it's running smoothly, failures on typescript@next should be fixed quickly.

**Comment by Flarna** [comment]
>  I'd rather not merge a change to a package as important as node without a green CI run unless there's something that makes it urgent.

Well, sooner or later this will result in loosing maintainers for node. It's frustrating if your approved PRs get closed by a bot again and again even all people (incl. typescript/dtslint experts/maintainers like @sanders) agree that CI fails are unrelated.

Here a short list of stalled PRs for node:

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/33019 - approved since 13 days, closed by bot because unrelated CI fails

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/33081 - approved since 12 days

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/33127 - approved since 9 days

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/33137 - approved since 9 days

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/33195 - approved since 7 days (this is already a followup as bot has closed ohters)

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/33176 - approved since 3 days

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/33325 - approved since 3 days

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/33356 - approved since a day

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/33045 - approved since 13 days
has been created as https://github.com/DefinitelyTyped/DefinitelyTyped/pull/31709 was closed by bot because of unrelated CI fails

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/33177 - approved since 8 days

**Comment by sandersn** [comment]
@Flarna Definitely Typed is healthy again. I also changed CI runs to only run the type-checking lint rule for dependent packages. That way at least other lint failures won't block node changes, and CI will complete a bit faster.

**Comment by Flarna** [comment]
@sandersn Cool, thanks a lot.

Could you please re-trigger CI for tickets mentioned [there](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/33127#issuecomment-467581533)? Or should we ping the authors to rebase?

**Comment by sandersn** [comment]
@Flarna there are some fixes in ember so the authors will need to rebase.

**Comment by Flarna** [comment]
@sandersn usually not as github tests against current master therefore you get fixes (and bugs) for free. So a close/re-open on them should greenify them.

**Comment by SimonSchick** [comment]
@Flarna @sandersn we good now? 😄

**Comment by Flarna** [comment]
@SimonSchick Any major changes since I approved?

**Comment by SimonSchick** [comment]
No, I just rebased.

**Comment by SimonSchick** [comment]
@jessetrinity maybe?

**Pull request merged**
